### ISSUE TITLE
Database error when special characters used in path

### ISSUE THREAD

**Issue by RyuMaster** [xaya/xaya#78]
```
2018-12-29T14:06:42Z BerkeleyEnvironment::Open: LogDir=C:\Users\Даниелька\AppData\Roaming\Xaya\wallets\vault.dat\database ErrorFile=C:\Users\Даниелька\AppData\Roaming\Xaya\wallets\vault.dat\db.log
2018-12-29T14:06:42Z BerkeleyEnvironment::Open: Error -30974 opening database environment: DB_RUNRECOVERY: Fatal error, run database recovery
```

Does not happen for latest namecoin or bitcoin client. Happens only when special characters are in path, tested on Windows 10 64x;

**Comment by domob1812** [comment]
How is this related to #79?  Just from the bug descriptions, it sounds like this bug is a direct consequence of #79 - or is the behaviour different (e.g. when the special characters are due to the default path vs in an explicit `-datadir`)?

**Comment by domob1812** [comment]
Also, when you say that this works for upstream Namecoin and Bitcoin:  Have you started those with the exact same configuration as Xaya Core, i.e. with `vault` and `game` wallets?  The error message above explicitly complains about the `vault` wallet file.  So perhaps this is something that only happens with a multi-wallet setting.

**Comment by RyuMaster** [comment]
This is not related to #79, hence I opened 2 seperate issues,
I did start those without explicitly naming wallets. Will do so and get back

**Comment by xaya** [comment]
Yes this is not related to multiwallets.

The same issue if you run the daemon without any options (Default datadir etc)

might be worth testing on Linux too

**Comment by domob1812** [comment]
@xaya:  As I stated in #79, the regtests actually verify that an explicit `-datadir` with special characters works fine.  This leaves us with two options:

1) This is broken on Windows only - please try running the regtests (`test/functional/test_runner.py`) on Windows, if possible.
2) This is broken only for the default datadir, not when explicitly setting it with `-datadir`.  Do you know already for sure whether this is the case or not?

**Comment by domob1812** [comment]
Also, what does the database error file (as listed in the error message from the original post) state?

**Comment by RyuMaster** [comment]
I ran daemon directly using this:

`xayad.exe -wallet=vault.dat  -wallet=game.dat -server -rpcallowip=127.0.0.1 -datadir="C:\ХахахаКостик2" -zmqpubhashtx=tcp://127.0.0.1:28332 -zmqpubhashblock=tcp://127.0.0.1:28332 -zmqpubrawblock=tcp://127.0.0.1:28332 -zmqpubrawtx=tcp://127.0.0.1:28332`

the C:\ХахахаКостик2\vault.dat\db.log error looks is as follows:

`xayad.exe -wallet=vault.dat  -wallet=game.dat -server -rpcallowip=127.0.0.1 -datadir="C:\ХахахаКостик2" -zmqpubhashtx=tcp://127.0.0.1:28332 -zmqpubhashblock=tcp://127.0.0.1:28332 -zmqpubrawblock=tcp://127.0.0.1:28332 -zmqpubrawtx=tcp://127.0.0.1:28332`

debug.log is typical:

```
2019-01-02T07:31:52Z Using wallet wallet.dat
2019-01-02T07:31:52Z BerkeleyEnvironment::Open: LogDir=C:\ХахахаКостик2\vault.dat\database ErrorFile=C:\ХахахаКостик2\vault.dat\db.log
2019-01-02T07:31:52Z BerkeleyEnvironment::Open: Error -30974 opening database environment: DB_RUNRECOVERY: Fatal error, run database recovery
2019-01-02T07:31:52Z Moved old C:\ХахахаКостик2\vault.dat\database to C:\ХахахаКостик2\vault.dat\database.1546414312.bak. Retrying.
2019-01-02T07:31:52Z BerkeleyEnvironment::Open: LogDir=C:\ХахахаКостик2\vault.dat\database ErrorFile=C:\ХахахаКостик2\vault.dat\db.log
2019-01-02T07:31:52Z BerkeleyEnvironment::Open: Error -30974 opening database environment: DB_RUNRECOVERY: Fatal error, run database recovery
2019-01-02T07:31:52Z Error: Error initializing wallet database environment "C:\ХахахаКостик2\vault.dat"!
2019-01-02T07:31:52Z Shutdown: In progress...
2019-01-02T07:31:52Z scheduler thread interrupt
2019-01-02T07:31:52Z Shutdown: done
```

**Comment by kr1z1s** [comment]
> I ran daemon directly using this:
> 
> `xayad.exe -wallet=vault.dat -wallet=game.dat -server -rpcallowip=127.0.0.1 -datadir="C:\ХахахаКостик2" -zmqpubhashtx=tcp://127.0.0.1:28332 -zmqpubhashblock=tcp://127.0.0.1:28332 -zmqpubrawblock=tcp://127.0.0.1:28332 -zmqpubrawtx=tcp://127.0.0.1:28332`
> 
> the C:\ХахахаКостик2\vault.dat\db.log error looks is as follows:
> 
> `C:\ХахахаКостик2\vault.dat: No such file or directory C:\ХахахаКостик2\vault.dat: No such file or directory C:\ХахахаКостик2\vault.dat\database: No such file or directory PANIC: No such file or directory process-private: unable to find environment C:\ХахахаКостик2\vault.dat: No such file or directory`
> 
> debug.log is typical:
> 
> ```
> 2019-01-02T07:31:52Z Using wallet wallet.dat
> 2019-01-02T07:31:52Z BerkeleyEnvironment::Open: LogDir=C:\ХахахаКостик2\vault.dat\database ErrorFile=C:\ХахахаКостик2\vault.dat\db.log
> 2019-01-02T07:31:52Z BerkeleyEnvironment::Open: Error -30974 opening database environment: DB_RUNRECOVERY: Fatal error, run database recovery
> 2019-01-02T07:31:52Z Moved old C:\ХахахаКостик2\vault.dat\database to C:\ХахахаКостик2\vault.dat\database.1546414312.bak. Retrying.
> 2019-01-02T07:31:52Z BerkeleyEnvironment::Open: LogDir=C:\ХахахаКостик2\vault.dat\database ErrorFile=C:\ХахахаКостик2\vault.dat\db.log
> 2019-01-02T07:31:52Z BerkeleyEnvironment::Open: Error -30974 opening database environment: DB_RUNRECOVERY: Fatal error, run database recovery
> 2019-01-02T07:31:52Z Error: Error initializing wallet database environment "C:\ХахахаКостик2\vault.dat"!
> 2019-01-02T07:31:52Z Shutdown: In progress...
> 2019-01-02T07:31:52Z scheduler thread interrupt
> 2019-01-02T07:31:52Z Shutdown: done
> ```

Do not use the Cyrillic alphabet.

**Comment by RyuMaster** [comment]
> Do not use the Cyrillic alphabet.
I do not, but some XAYA Electron wallet users have it inside their default path as windows username

**Comment by RyuMaster** [comment]
> Thanks. To me that looks like some issue with BDB, but then I wonder why it works for Bitcoin. Have you tried whether that happens only with the default datadir (username contains cyrillic characters), or whether it also happens on Windows if you use an explicit `-datadir`?

In the example logs above, you can see that I am passing datadir explicit. 

Yes, this is odd, that it works for bitcoin. I also tested namecoin with the exactly same configuration options as in the example above, and it worked too.

**Comment by RyuMaster** [comment]
> Can you confirm that test/functional/test_runner.py fails for that reason on Windows?

I did try, but  test/functional/test_runner.py  gives me all kind of trouble even on normal English windows version. Maybe it was not tested with win properly, or I need some specific python configuration to run it

**Comment by domob1812** [comment]
> > Can you confirm that test/functional/test_runner.py fails for that reason on Windows?
> 
> I did try, but test/functional/test_runner.py gives me all kind of trouble even on normal English windows version. Maybe it was not tested with win properly, or I need some specific python configuration to run it

For Windows, it is certainly tested in upstream Bitcoin.  But I don't know if it works with cross-compile, and if it needs special attention to run.  Then don't worry about it for now.

**Comment by xaya** [comment]
this issue is now fixed upstream here > https://github.com/bitcoin/bitcoin/issues/15089

but another related ish issue here (although it is a separate problem still involves special characters) > https://github.com/bitcoin/bitcoin/issues/15092

**Comment by xaya** [comment]
I have built and tested xaya 1.1 branch with the bdb fix https://github.com/bitcoin/bitcoin/issues/15089 + leveldb fix here https://github.com/bitcoin/bitcoin/issues/15092
for windows 64bit and 32bit and it seems to fix both issues.

@domob1812 I can release these windows binaries but are you able to go over both those fixes to see if they could cause any major issues anywhere else?
we can then wait until they are applied upstream (first one has been afaik) which will hopefully be done before the next major update (probably hardfork) with XAYA?

unless you think we should merge those fixes to xaya before hand?

either way we need to provide the fixed binaries as soon as we can.

**Comment by domob1812** [comment]
I'm not an expert about the touched code parts, but as far as I can tell, yes, it should be fine to release fixed binaries for now.  I also agree that we should wait for those changes to be merged upstream for the main codebase.  (And yes, looks like the first patch is indeed already merged upstream - so will be merged to Xaya soon, e.g. next week.)

**Comment by domob1812** [comment]
Also the LevelDB fix is now merged upstream and in Xaya's `master`.  Both fixes are also backported to `1.1`.
